### ISSUE TITLE
error: ambiguous deduction for template arguments of 'Ok'

### ISSUE THREAD

**Issue by makotokato** [unicode-org/icu4x#4109]
When building `icu_capi`'s header with clang version 17.0.1 (+ a few customization for mozilla), the following error occurs.  Known issue?

```
 4:49.40 /mozilla/mozilla-central/third_party/rust/icu_capi/cpp/include/ICU4XLocale.hpp:244:33: error: ambiguous deduction for template arguments of 'Ok'
 4:49.40   244 |     diplomat_result_out_value = diplomat::Ok(std::monostate());
 4:49.40       |                                 ^
 4:49.40 /mozilla/mozilla-central/third_party/rust/icu_capi/cpp/include/diplomat_runtime.hpp:57:3: note: candidate function [with T = std::monostate]
 4:49.40    57 |   Ok(T&& i): inner(std::move(i)) {}
 4:49.40       |   ^
 4:49.40 /mozilla/mozilla-central/third_party/rust/icu_capi/cpp/include/diplomat_runtime.hpp:61:3: note: candidate function [with T = std::monostate, X = std::monostate, $2 = void]
 4:49.40    61 |   Ok(T i): inner(i) {}
```

**Comment by sffc** [comment]
Note that the C++ diplomat headers changed slightly here

https://github.com/rust-diplomat/diplomat/pull/336

Can you make sure you are using the latest version of all diplomat header files?

**Comment by sffc** [comment]
CC @Manishearth

**Comment by Manishearth** [comment]
Yeah, I'm not too clear on C++ templates to know what's going on here.

https://github.com/rust-diplomat/diplomat/pull/336/files#diff-56a0cb4d0a6aa11948a874d240f37965752450afd7bcb6a748fcb535f121cd40

^ that's the relevant change; I think we made that after talking to a C++ expert but if you can find a C++ expert to provide better direction here I'd be happy to update things.

**Comment by sffc** [comment]
@makotokato The change here was intended to resolve the issue you had previously filed, https://github.com/rust-diplomat/diplomat/issues/335. The change passes in CI. We do not understand your build environment so would appreciate if you could help suggest a fix for the issue, which we can land upstream in Diplomat.

**Comment by Manishearth** [comment]
Ideally it would be nice if you could contact C++ experts at Mozilla and help resolve this. In general ambiguity between move and normal constructors hasn't been a thing in my experience.

**Comment by robertbastian** [comment]
I can reproduce these errors with `cargo make test-cpp-emscripten`.

**Comment by makotokato** [comment]
https://github.com/llvm/llvm-project/issues/70162

**Comment by Manishearth** [comment]
Ah, not our fault perhaps? Still worth mitigating if possible.

**Comment by Manishearth** [comment]
https://github.com/rust-diplomat/diplomat/pull/429 should fix this.

The reason we weren't catching this on CI is that this is a GCC/Clang issue and we CI on GCC.

@sffc I'd like to backport this patch.  This probably involves:

 - Releasing a patch release of diplomat-tool 0.7.1, with the backport. The API has changed so we can't release main unless we release it as 0.8. Which is also acceptable. Or we release nothing and use a locally-generated icu_capi in the patch release. I'm not sure what to prefer
 - Releasing a patch release of icu_capi.

**Comment by Manishearth** [comment]
@makotokato We'll be doing a backport that will make this available in ICU4X 1.4. https://github.com/unicode-org/icu4x/pull/4616
